### ISSUE TITLE
The failure of the segmentation

### ISSUE THREAD

**Issue by ChenFeiL** [Kamnitsask/deepmedic#19]
Hi ,

I have been trying to work with the deepmedic package on the Brats2015. I downloaded them from the official website. All the processes with the deepmedic network run absolutely fine. However, after the testing process , all the pixel values of the files named "pred_brats_xxx_hgg_Segm.nii" are 0. I'm copying the file below for your reference. I would appreciate your help on this.

-ChenFeiL
[pred_brats_tcia_pat410_0001_hgg_ProbMapClass4.nii.gz](https://github.com/Kamnitsask/deepmedic/files/453773/pred_brats_tcia_pat410_0001_hgg_ProbMapClass4.nii.gz)

**Comment by Kamnitsask** [comment]
I will need some more information on exactly what you did.

Prior to training the network, did you normalize each images by getting the mean and std of the tissue within the brainmask? and normalize by subtracting mean, dividing std?

If you ran it on the data as you acquired them from the site without normalization, it wont work.  Configured parameters are set for normalized intensities.

I ll need details on exactly what you did.

K

**Comment by ChenFeiL** [comment]
Hi,
I am not very clear the normalization process here. I should getting the mean and std of the tissue within the brainmask, not the whole image? Then I should normalize the whole image by subtracting mean, dividing std. Is that right? I got the mean and std within the whole image before, and I met the mistake.

What's more, when I normalize the image within the brainmask, the pixel values are not the same with your   examples.

**Comment by Kamnitsask** [comment]
Hello again,

- Get the mean and std only within the brain.
- Works even better if you "window" the intensities a bit, eg exclude bottom and top 1% percent.
- Then for each voxel in the image, do : newIntensity = (intensity-mean) / std
Note: What you really want is that the intensities of the tissue that is going to be processed to be nicely spreaded with 0mean 1 std. So when you normalized with mean/std of whole image, which has loads of air, you squeezed the distribution a lot, and the within-tissue intensities were brought very close to each other.

Important:
Also remember to provide the ROI-Brainmask for training, so that sampling is only done within it, and not from the air as well.

Unfortunately, to my experience, CNNs are very sensitive to the preprocessing. So, feel free to let me know how this goes and ask anything further.

Kostas

**Comment by ChenFeiL** [comment]
Hi,

I'm sorry to reply so late. Thank you for your advice. I have solved my problem. Thank you !

-ChenFeiL

**Comment by Kamnitsask** [comment]
Can I ask what was the problem? For future reference?

**Comment by ChenFeiL** [comment]
Yes, it's the wrong normalization process that leads to the error. I get the mean and std only within the brain just as you said and this time I get the right results.

**Comment by Kamnitsask** [comment]
Thanks ChenFeil. 

Yes, the particular provided configuration of metaparameters is set for this type of normalization, and it seems very sensitive to it. I am sure that there must be better ways to normalize, or more convenient. But then one would need to adjust some hyperparameters,  mostly learning rate, to make it work in the new setting.

Thanks for the feedback, hope this software helps you somehow in your research. Good luck.

Kostas

**Comment by ChenFeiL** [comment]
Hi,

Bother you again. After the training process, I get the Dice Coefficient: 
DICE1=[0.8355 294.3985 0.6915 0.1884 294.6810]. There are 5 columns, what does the five columns mean?

-ChenFeiL

**Comment by Kamnitsask** [comment]
Dice score, one element for each label/class: [Class-0, class-1, class-2, ... ] 

This 294 is there cause some subjects have no such class at all, and the calculation goes wrong. You are using an older version I guess. This is fixed in the more recent versions. Repull the master.

Kostas
